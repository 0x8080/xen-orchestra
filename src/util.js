import {Xo} from 'xo-lib'
import {find, forEach, map, cloneDeep} from 'lodash'
import expect from 'must'

export async function getConfig () {
  return {
    adminCredentials: {
      email: 'admin@admin.net',
      password: 'admin'
    },
    xoServerUrl: 'localhost:9000',
    xenServer1: {
      host: '192.168.1.3',
      username: 'root',
      password: 'qwerty'
    }
  }
}

export async function getConnection ({
  credentials
} = {}) {
  const config = await getConfig()
  const xo = new Xo(config.xoServerUrl)
  await xo.signIn(
    credentials === undefined ?
      config.adminCredentials :
      credentials
  )

  // Injects waitObject()
  //
  // TODO: integrate in xo-lib.
  const watchers = {}
  xo.waitObject = (id) => {
    return new Promise(resolve => {
      watchers[id] = resolve
    })
  }

  const onUpdate = objects => {
    forEach(objects, (object, id) => {
      if (watchers[id]) {
        watchers[id](object)
        delete watchers[id]
      }
    })
  }
  xo.objects.on('add', onUpdate)
  xo.objects.on('update', onUpdate)
  xo.objects.on('remove', onUpdate)

  xo.getOrWaitObject = async function (id) {
    const object = this.objects.all[id]
    if (object) {
      return object
    }

    return await this.waitObject(id)
  }

  return xo
}

// =================================================================

export async function getAllUsers (xo) {
  return await xo.call('user.getAll')
}

export async function getUser (xo, id) {
  const users = await getAllUsers(xo)
  return find(users, {id: id})
}

export async function createUser (xo, userIds, params) {
  const userId = await xo.call('user.create', params)
  userIds.push(userId)
  return userId
}

export async function deleteUsers (xo, userIds) {
  await Promise.all(map(
    userIds,
    userId => xo.call('user.delete', {id: userId})
  ))
}

// ==================================================================

export function getAllHosts (xo) {
  return xo.objects.indexes.type.host
}

export function getOneHost (xo) {
  const hosts = getAllHosts(xo)
  for (const id in hosts) {
    return hosts[id]
  }

  throw new Error('no hosts found')
}

// ==================================================================

export function deepDelete (obj, path) {
  const lastIndex = path.length - 1
  for (let i = 0; i < lastIndex; i++) {
    obj = obj[path[i]]

    if (typeof obj !== 'object' || obj === null) {
      return
    }
  }
  delete obj[path[lastIndex]]
}

export function almostEqual (actual, expected, ignoredAttributes) {
  actual = cloneDeep(actual)
  expected = cloneDeep(expected)
  forEach(ignoredAttributes, (ignoredAttribute) => {
    deepDelete(actual, ignoredAttribute.split('.'))
    deepDelete(expected, ignoredAttribute.split('.'))
  })
  expect(actual).to.be.eql(expected)
}

export async function waitObjectState (xo, id, predicate) {
  let obj = xo.objects.all[id]
  while (true) {
    try {
      await predicate(obj)
      return
    } catch (_) {}
    // If failed, wait for next object state/update and retry.
    obj = await xo.waitObject(id)
  }
}
