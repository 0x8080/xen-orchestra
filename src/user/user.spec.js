/* eslint-env jest */

import { forOwn, keyBy, omit } from "lodash";
import { testConnection, xo } from "../util";

const simpleUser = {
  email: "wayne3@vates.fr",
  password: "batman",
};

const withData = (data, fn) =>
  forOwn(data, (data, title) => {
    it(title, () => fn(data));
  });

describe("user", () => {
  describe(".create() :", () => {
    withData(
      {
        "creates a user without permission": {
          email: "wayne1@vates.fr",
          password: "batman1",
        },
        "creates a user with permission": {
          email: "wayne2@vates.fr",
          password: "batman2",
          permission: "user",
        },
      },
      async data => {
        const userId = await xo.createUser(data);
        expect(typeof userId).toBe("string");
        expect(omit(await xo.getUser(userId), "id")).toMatchSnapshot();
        await testConnection({
          credentials: {
            email: data.email,
            password: data.password,
          },
        });
      }
    );

    withData(
      {
        "fails trying to create a user without email": { password: "batman" },
        "fails trying to create a user without password": {
          email: "wayne@vates.fr",
        },
      },
      async data => {
        await expect(xo.createUser(data)).rejects.toMatchSnapshot();
      }
    );

    it("fails trying to create a user with an email already used", async () => {
      await xo.createUser(simpleUser);
      await expect(xo.createUser(simpleUser)).rejects.toMatchSnapshot();
    });
  });

  describe(".getAll() :", () => {
    it("gets all the users created", async () => {
      const userId1 = await xo.createUser({
        email: "wayne4@vates.fr",
        password: "batman",
        permission: "user",
      });
      const userId2 = await xo.createUser({
        email: "wayne5@vates.fr",
        password: "batman",
        permission: "user",
      });
      let users = await xo.call("user.getAll");
      expect(Array.isArray(users)).toBe(true);
      users = keyBy(users, "id");
      expect([
        omit(users[userId1], "id"),
        omit(users[userId2], "id"),
      ]).toMatchSnapshot();
    });
  });

  describe(".delete() :", () => {
    it("deletes a user successfully with id", async () => {
      const userId = await xo.call("user.create", simpleUser);
      expect(await xo.call("user.delete", { id: userId })).toBe(true);
      expect(await xo.getUser(userId)).toBe(undefined);
    });

    it("fails trying to delete a user with a nonexistent user", async () => {
      await expect(
        xo.call("user.delete", { id: "nonexistentId" })
      ).rejects.toMatchSnapshot();
    });

    it("fails trying to delete itself", async () => {
      await expect(
        xo.call("user.delete", { id: xo._user.id })
      ).rejects.toMatchSnapshot();
    });
  });
});
