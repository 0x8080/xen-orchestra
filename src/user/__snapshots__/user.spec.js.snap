// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user .create() : creates a user with permission 1`] = `
Object {
  "email": "wayne2@vates.fr",
  "groups": Array [],
  "permission": "user",
  "preferences": Object {},
}
`;

exports[`user .create() : creates a user without permission 1`] = `
Object {
  "email": "wayne1@vates.fr",
  "groups": Array [],
  "preferences": Object {},
}
`;

exports[`user .create() : fails trying to create a user with an email already used 1`] = `[JsonRpcError: unknown error from the peer]`;

exports[`user .create() : fails trying to create a user without email 1`] = `[JsonRpcError: invalid parameters]`;

exports[`user .create() : fails trying to create a user without password 1`] = `[JsonRpcError: invalid parameters]`;

exports[`user .delete() : fails trying to delete a user with a nonexistent user 1`] = `[JsonRpcError: no such user nonexistentId]`;

exports[`user .delete() : fails trying to delete itself 1`] = `[JsonRpcError: a user cannot delete itself]`;

exports[`user .getAll() : gets all the users created 1`] = `
Array [
  Object {
    "email": "wayne4@vates.fr",
    "groups": Array [],
    "permission": "user",
    "preferences": Object {},
  },
  Object {
    "email": "wayne5@vates.fr",
    "groups": Array [],
    "permission": "user",
    "preferences": Object {},
  },
]
`;
